File: cs/ByteComm/ByteServer.cs
using System;
using System.Text;
using NetMQ;
using NetMQ.Sockets;

namespace ByteComm
{
    public class ByteServer
    {
        readonly ResponseSocket socket;

        public ByteServer(string sockname)
        {
            socket = new ResponseSocket();
            socket.Bind(sockname);
        }

        public void Send_Byte(byte[] message)
        {
            socket.SendFrame(message);
        }

        public void Send_String(string message)
        {
            socket.SendFrame(Encoding.ASCII.GetBytes(message));
        }

        public byte[] Recv_Byte()
        {
            return socket.ReceiveFrameBytes();
        }

        public string Recv_String()
        {
            return socket.ReceiveFrameString();
        }

        public string Try_Recv_String()
        {
            string a;
            bool result = socket.TryReceiveFrameString(out a);
            return result? a: "";
        }

        public void Dummy_Serve()
        {
            while (true)
            {
                string msg_recv = Recv_String();
                string msg = "From C#, I've got: " + msg_recv;
                Send_String(msg);
            }
        }

        public void Close()
        {
            socket.Close();
        }
    }
}

File: ruby/ByteServer.rb
require './Util'

module CrowdComm
  require 'ffi-rzmq'

  class ByteServer
    include ZMQ
    include CrowdComm::Utils

    def initialize(sockname)
      @context = Context.new(1)
      @pipe = @context.socket(REP)
      if got_error @pipe.bind(sockname)
        raise "Fail to listen on #{sockname}" 
      end
      puts "Listening on #{sockname}"
    end

    def send_bytes(bytes)
      if got_error @pipe.send_string(bytes)
        raise 'Fail to send message'
      end
    end

    def get_bytes
      msg = ''
      if got_error @pipe.recv_string(msg)
        raise 'Fail to receive message'
      end
      msg
    end

    def dummy_serve
      byte_decode = proc do |b|
        b.to_i
      end

      begin
        loop do
          msg = get_bytes
          send_bytes "I've got: #{msg}"
        end
      rescue Interrupt => e
        puts "\rServer shutting down..."
        @pipe.close
      end
    end
  end
end

if __FILE__ == $0
  CrowdComm::ByteServer.new('ipc:///tmp/sock_temp').dummy_serve
end
